// Copyright 2020-present Open Networking Foundation
// SPDX-License-Identifier: Apache-2.0

syntax = "proto3";

option cc_generic_services = false;

package stratum.hal.barefoot;

import "p4/config/v1/p4info.proto";

// This message describes the public Barefoot specific definition of a single P4
// program. It is passed to Stratum as the p4 device config of the P4Runtime
// ForwardingPipelineConfig message.
message BfPipelineConfig {
  message Profile {
    string profile_name = 1;
    bytes context = 2;
    bytes binary = 3;
    repeated uint32 pipe_scope = 4;
  }
  string p4_name = 1;
  bytes bfruntime_info = 2;
  repeated Profile profiles = 3;
}

// Internal storage format that contains all the information of a P4 program
// that can be loaded into the SDE.
message BfrtDeviceConfig {
  message P4Program {
    message P4Pipeline {
      string name = 1;           // For display purposes only.
      repeated uint32 scope = 2; // List of pipes this profile is applied to.
      bytes context = 3;         // JSON
      bytes config = 4;          // tofino.bin
    }

    string name = 1; // For display purposes only.
    bytes bfrt = 2;  // JSON
    p4.config.v1.P4Info p4info = 3;
    repeated P4Pipeline pipelines = 4;
  }

  repeated P4Program programs = 1;
}
